Replace duplicate logic with equivalent SymbolKit API

heckj opened the pull request:
Originally found while debugging #1084, and recently updated the logic in https://github.com/swiftlang/swift-docc-symbolkit/pull/101 to match this version, which was used in preference of the SymbolKit version.

## Summary

This removes the logic from DocC, favoring the loading in SymbolKit, and consolidating the logic into a single place. This was originally opened at #1091, but the code had changed significantly since that update, so it was easier to create a new PR with the change after applying feedback from the earlier PR.

The equivalent logic that I'm consolidating towards is https://github.com/swiftlang/swift-docc-symbolkit/blob/main/Sources/SymbolKit/UnifiedSymbolGraph/GraphCollector.swift#L121-L178

## Testing

The existing tests within DocC work this logic extensively.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests - _no additional tests needed_
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary - _no documentation changes are relevant, internal implementation details only_

heckj committed:
remove shadowed method that's duplicated in SymbolKit, and consistently use the SymbolKit version

d-ronnqvist commented:
If the SymbolKit logic for this was recently updated, does it make sense to also run `swift package update swift-docc-symbolkit ` to update the commit of SymbolKit that DocC depends on?

heckj committed:
consistently use SymbolKit GraphCollector.moduleNameFor in PathHierarchy too

Removes the last DocC-local duplicate of this logic, the file-name-only moduleNameFor(_:URL) method, by using the SymbolKit implementation at its one remaining call site. Same idea as swift-docc#1397, without the parts of that PR that depend on unmerged upstream changes.

snprajwal reviewed (approved):
LGTM, thank you!

snprajwal committed:
Merge branch 'main' into remove-duplicate-logic
